lib/std/simd.zig

```diff
@@ -0,0 +1,395 @@
+    const CountIntType = VectorCount(@TypeOf(vec));
+
+    const one_if_true = @select(CountIntType, vec, @splat(len, @as(CountIntType, 1)), @splat(len, @as(CountIntType, 0)));
+    return @reduce(.Add, one_if_true);
+}
+
+pub fn firstIndexOfValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return firstTrue(vec == @splat(len, value));
+}
+
+pub fn lastIndexOfValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return lastTrue(vec == @splat(len, value));
+}
+
+pub fn countElementsWithValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) VectorCount(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return countTrues(vec == @splat(len, value));
+}
+
+test "vector searching" {
+    const base = Vector(8, u32){ 6, 4, 7, 4, 4, 2, 3, 7 };
+
+    try std.testing.expectEqual(@as(?u3, 1), firstIndexOfValue(base, 4));
+    try std.testing.expectEqual(@as(?u3, 4), lastIndexOfValue(base, 4));
+    try std.testing.expectEqual(@as(?u3, null), lastIndexOfValue(base, 99));
+    try std.testing.expectEqual(@as(u4, 3), countElementsWithValue(base, 4));
+}
+
+/// Same as accum, but with a user-provided function that mathematically must be associative.
+/// The function must also do nothing when called with identity as either argument.
+/// Set ErrorType to void if func doesn't return an error union.
+pub fn accumWithFunc(
+    comptime hop: isize,
+    vec: anytype,
+    comptime ErrorType: type,
+    comptime func: fn (@TypeOf(vec), @TypeOf(vec)) if (ErrorType == void) @TypeOf(vec) else ErrorType!@TypeOf(vec),
+    comptime identity: std.meta.Child(@TypeOf(vec)),
+) if (ErrorType == void) @TypeOf(vec) else ErrorType!@TypeOf(vec) {
+    // I haven't debugged this, but it might be a cousin of sorts to what's going on with braid.
+    comptime if (builtin.cpu.arch.isMIPS()) @compileError("TODO: Find out why accum doesn't work on MIPS");
+
+    const len = vectorLength(@TypeOf(vec));
+
+    if (hop == 0) @compileError("hop can not be 0; you'd be going nowhere forever!");
+    const abs_hop = if (hop < 0) -hop else hop;
+
+    var acc = vec;
+    comptime var i = 0;
+    inline while ((abs_hop << i) < len) : (i += 1) {
+        const shifted = if (hop < 0) shiftElementsDown(acc, abs_hop << i, identity) else shiftElementsUp(acc, abs_hop << i, identity);
+
+        acc = if (ErrorType == void) func(acc, shifted) else try func(acc, shifted);
+    }
+    return acc;
+}
+
+/// Returns a vector whose elements are the result of performing the specified operation on the corresponding
+/// element of the input vector and every hop'th element that came before it (or after, if hop is negative).
+/// Supports the same operations as the @reduce() builtin. Takes O(logN) to compute.
+pub fn accum(comptime op: std.builtin.ReduceOp, comptime hop: isize, vec: anytype) @TypeOf(vec) {
```

Comment:
This operation is usually either called prefix sum (for addition) or prefix scan (generic). Its also worth noting here whether the operation is inclusive or exclusive.